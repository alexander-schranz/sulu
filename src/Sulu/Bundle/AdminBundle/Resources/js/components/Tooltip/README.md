This component displays a tooltip if its content is hovered or focused:

```javascript
import Icon from '../Icon';

<Tooltip label="Some hover text" aria-label="Some hover text">
    Hover to see the tooltip
</Tooltip>
```

The tooltip itself is rendered on the root body element to avoid z-index problems, as this means
it is out of context it is rendered as `aria-hidden`.

For accessibility reasons your button or element should define the `aria-label` attribute.

A button with a tooltip should be wrapped inside the tooltip so focus on the button will also
show the tooltip:

```javascript
import Icon from '../Icon';
const buttonStyle = {
    background: 'none',
    border: 'none',
    cursor: 'pointer',
};

<div style={{display: 'flex', gap: '20px'}}>
    <Tooltip label="Copy">
        <button type="button" onClick={() => {alert("Copy")}} aria-label="Copy" style={buttonStyle}>
            <Icon name="su-copy" />
        </button>
    </Tooltip>

    <Tooltip label="Duplicate">
        <button type="button" onClick={() => {alert("Duplicate")}} aria-label="Duplicate" style={buttonStyle}>
            <Icon name="su-duplicate" />
        </button>
    </Tooltip>

    <Tooltip label="Cut">
        <button type="button" onClick={() => {alert("Cut")}} aria-label="Cut" style={buttonStyle}>
            <Icon name="su-cut" />
        </button>
    </Tooltip>

    <Tooltip label="Delete">
        <button type="button" onClick={() => {alert("Delete")}} aria-label="Delete" style={buttonStyle}>
            <Icon name="su-trash-alt" />
        </button>
    </Tooltip>
</div>
```
