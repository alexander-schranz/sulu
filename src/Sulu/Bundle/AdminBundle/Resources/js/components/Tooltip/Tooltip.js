// @flow
import React from 'react';
import classNames from 'classnames';
import {action, observable} from 'mobx';
import {observer} from 'mobx-react';
import Popover from '../Popover';
import tooltipStyles from './tooltip.scss';
import type {ChildrenArray, Element, ElementRef} from 'react';

type Props = {|
    children: ChildrenArray<Element<*> | false>,
    label: string,
|};

@observer
class Tooltip extends React.Component<Props> {
    constructor(props: Props) {
        super(props);
    }

    @observable tooltipOpen: boolean = false;

    @observable tooltipRef: ?ElementRef<*>;

    @action setTooltipRef = (ref: ?ElementRef<*>) => {
        this.tooltipRef = ref;
    };

    @action handleEnter = () => {
        this.tooltipOpen = true;
    };

    @action handleLeave = () => {
        this.tooltipOpen = false;
    };

    render() {
        const {
            children,
            label,
        } = this.props;

        return (
            // eslint-disable-next-line jsx-a11y/no-static-element-interactions
            <span
                className={tooltipStyles.tooltipContainer}
                onBlur={this.handleLeave}
                onFocus={this.handleEnter}
                onMouseEnter={this.handleEnter}
                onMouseLeave={this.handleLeave}
                ref={this.setTooltipRef}
            >
                {
                    this.tooltipRef
                        && <Popover
                            anchorElement={this.tooltipRef}
                            backdrop={false}
                            horizontalCenter={true}
                            open={this.tooltipOpen}
                            verticalOffset={10}
                        >
                            {
                                (setPopoverRef, styles, verticalPosition) => (
                                    <span
                                        aria-hidden={true}
                                        className={classNames(tooltipStyles.tooltip, tooltipStyles[verticalPosition])}
                                        ref={setPopoverRef}
                                        style={styles}
                                    >
                                        {label}
                                    </span>
                                )
                            }
                        </Popover>
                }

                {children}
            </span>
        );
    }
}

export default Tooltip;
