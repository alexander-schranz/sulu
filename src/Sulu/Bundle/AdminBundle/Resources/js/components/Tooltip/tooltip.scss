@import '../../containers/Application/colors.scss';

.tooltipContainer {
    display: inline-block;
    background: none;
    border: none;
}

.tooltip {
    background: $black;
    border-radius: 3px;
    color: $white;
    font-weight: 600;
    font-size: 10px;
    line-height: 14px;
    padding: 3px 7px 3px;
}

.top {
    position: relative;

    &::before {
        content: '';
        position: absolute;
        top: 100%;
        left: 50%;
        margin-left: -6px;
        border-width: 4px 6px 0 6px;
        border-style: solid;
        border-color: $black transparent transparent transparent;
    }
}

.bottom {
    position: relative;

    &::before {
        content: '';
        position: absolute;
        top: -4px;
        left: 50%;
        margin-left: -6px;
        border-width: 0 6px 4px 6px;
        border-style: solid;
        border-color: transparent transparent $black transparent;
    }
}
