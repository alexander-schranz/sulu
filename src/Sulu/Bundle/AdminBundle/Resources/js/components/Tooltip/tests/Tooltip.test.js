// @flow
import React from 'react';
import {mount, render} from 'enzyme';
import Tooltip from '../Tooltip';
import Icon from '../../Icon';

test('The component should render in unfocused state', () => {
    const component = render(
        <Tooltip label="Copy">
            <button aria-label="Copy" type="button">
                <Icon name="su-copy" />
            </button>
        </Tooltip>
    );

    expect(component.find('Popover span').length).toBe(0);

    expect(component).toMatchSnapshot();
});

test('The component should render in focused state', () => {
    const component = mount(
        <Tooltip label="Copy">
            <button aria-label="Copy" type="button">
                <Icon name="su-copy" />
            </button>
        </Tooltip>
    );

    component.find('button').simulate('focus');

    expect(component.find('Popover span').text()).toBe('Copy');
    expect(component).toMatchSnapshot();
});

test('The component should render in hovered state', () => {
    const component = mount(
        <Tooltip label="Copy">
            <button aria-label="Copy" type="button">
                <Icon name="su-copy" />
            </button>
        </Tooltip>
    );

    component.find('Tooltip').simulate('mouseenter');

    expect(component.find('Popover span').text()).toBe('Copy');
    expect(component).toMatchSnapshot();

    component.find('Tooltip').simulate('mouseleave');
    expect(component.find('Popover span').length).toBe(0);
});
