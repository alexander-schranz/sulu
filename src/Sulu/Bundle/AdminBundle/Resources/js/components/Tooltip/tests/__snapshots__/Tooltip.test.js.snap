// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should render in focused state 1`] = `
<Tooltip
  label="Copy"
>
  <span
    className="tooltipContainer"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <Popover
      anchorElement={
        <span
          class="tooltipContainer"
        >
          <button
            aria-label="Copy"
            type="button"
          >
            <span
              aria-label="su-copy"
              class="su-copy"
            />
          </button>
        </span>
      }
      backdrop={false}
      horizontalAnchorMode="center"
      horizontalOffset={0}
      open={true}
      verticalOffset={10}
    >
      <Portal>
        <Portal
          containerInfo={
            <div>
              <div
                class="container"
              >
                <span
                  aria-hidden="true"
                  class="tooltip bottom"
                  style="top: 10px; position: fixed; pointer-events: auto; left: 10px;"
                >
                  Copy
                </span>
              </div>
            </div>
          }
        >
          <div
            className="container"
          >
            <span
              aria-hidden={true}
              className="tooltip bottom"
              style={
                Object {
                  "left": "10px",
                  "maxHeight": undefined,
                  "pointerEvents": "auto",
                  "position": "fixed",
                  "top": "10px",
                }
              }
            >
              Copy
            </span>
          </div>
        </Portal>
      </Portal>
    </Popover>
    <button
      aria-label="Copy"
      type="button"
    >
      <Icon
        name="su-copy"
      >
        <span
          aria-label="su-copy"
          className="su-copy"
        />
      </Icon>
    </button>
  </span>
</Tooltip>
`;

exports[`The component should render in hovered state 1`] = `
<Tooltip
  label="Copy"
>
  <span
    className="tooltipContainer"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <Popover
      anchorElement={
        <span
          class="tooltipContainer"
        >
          <button
            aria-label="Copy"
            type="button"
          >
            <span
              aria-label="su-copy"
              class="su-copy"
            />
          </button>
        </span>
      }
      backdrop={false}
      horizontalAnchorMode="center"
      horizontalOffset={0}
      open={true}
      verticalOffset={10}
    >
      <Portal>
        <Portal
          containerInfo={
            <div>
              <div
                class="container"
              >
                <span
                  aria-hidden="true"
                  class="tooltip bottom"
                  style="top: 10px; position: fixed; pointer-events: auto; left: 10px;"
                >
                  Copy
                </span>
              </div>
            </div>
          }
        >
          <div
            className="container"
          >
            <span
              aria-hidden={true}
              className="tooltip bottom"
              style={
                Object {
                  "left": "10px",
                  "maxHeight": undefined,
                  "pointerEvents": "auto",
                  "position": "fixed",
                  "top": "10px",
                }
              }
            >
              Copy
            </span>
          </div>
        </Portal>
      </Portal>
    </Popover>
    <button
      aria-label="Copy"
      type="button"
    >
      <Icon
        name="su-copy"
      >
        <span
          aria-label="su-copy"
          className="su-copy"
        />
      </Icon>
    </button>
  </span>
</Tooltip>
`;

exports[`The component should render in unfocused state 1`] = `
<span
  class="tooltipContainer"
>
  <button
    aria-label="Copy"
    type="button"
  >
    <span
      aria-label="su-copy"
      class="su-copy"
    />
  </button>
</span>
`;
